test/integration/network_extension_discovery_integration_test.cc

```diff
@@ -0,0 +1,801 @@
+  addFilterChain();
+  addStaticFilter("bar", 2);
+  addStaticFilter("baz", 2);
+  addDynamicFilter(filter_name_, true);
+  addDynamicFilter(filter_name_, false);
+  initialize();
+
+  test_server_->waitForCounterGe("listener_manager.lds.update_success", 1);
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initializing);
+
+  sendXdsResponse(filter_name_, "1", 2);
+  test_server_->waitForCounterGe(
+      "extension_config_discovery.network_filter." + filter_name_ + ".config_reload", 1);
+  // filter drain 2 + 2 + 2 + 2 bytes.
+  sendDataVerifyResults(8);
+}
+
+TEST_P(NetworkExtensionDiscoveryIntegrationTest, DestroyDuringInit) {
+  // If rate limiting is enabled on the config source, gRPC mux drainage updates the requests
+  // queue size on destruction. The update calls out to stats scope nested under the extension
+  // config subscription stats scope. This test verifies that the stats scope outlasts the gRPC
+  // subscription.
+  on_server_init_function_ = [&]() { waitXdsStream(); };
+  addFilterChain();
+  addDynamicFilter(filter_name_, false, true);
+  initialize();
+
+  test_server_->waitForCounterGe("listener_manager.lds.update_success", 1);
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initializing);
+
+  test_server_.reset();
+  auto result = ecds_connection_->waitForDisconnect();
+  ASSERT_TRUE(result);
+  ecds_connection_.reset();
+}
+
+// Validate that a network filter update should fail if the subscribed extension configuration make
+// filter terminal but the filter position is not at the last position at filter chain. There would
+// be total of 2 filters in the chain: 'foo' and 'tcp_proxy' and both are marked terminal.
+TEST_P(NetworkExtensionDiscoveryIntegrationTest, BasicFailTerminalFilterNotAtEndOfFilterChain) {
+  on_server_init_function_ = [&]() { waitXdsStream(); };
+  addFilterChain();
+  addDynamicFilter(filter_name_, false, false);
+  initialize();
+
+  test_server_->waitForCounterGe("listener_manager.lds.update_success", 1);
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initializing);
+
+  sendXdsResponse(filter_name_, "1", 5, false, false, true);
+  test_server_->waitForCounterGe(
+      "extension_config_discovery.network_filter." + filter_name_ + ".config_fail", 1);
+  test_server_->waitUntilListenersReady();
+  test_server_->waitForGaugeGe("listener_manager.workers_started", 1);
+
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initialized);
+
+  // New connections will close since there's no valid configuration.
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort(port_name_));
+  auto result = tcp_client->write(data_);
+  if (result) {
+    tcp_client->waitForDisconnect();
+  }
+}
+
+// Basic ECDS config dump test with one filter.
+TEST_P(NetworkExtensionDiscoveryIntegrationTest, BasicSuccessWithConfigDump) {
```

Comment:
Thanks for config dump test.

It's probably valuable to add a network ecds filter config dump test here:

https://github.com/envoyproxy/envoy/blob/06e3e7d59fddc6541cecf399a8adc64092df7ed8/test/server/admin/config_dump_handler_test.cc#L809

Reply:
Thanks for the insight, added.